src/VecSim/spaces/L2/L2_NEON_FP16.h

```diff
@@ -0,0 +1,97 @@
+/*
+ *Copyright Redis Ltd. 2021 - present
+ *Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ *the Server Side Public License v1 (SSPLv1).
+ */
+
+#include <arm_neon.h>
+
+inline void L2Sqr_Step(const float16_t *&vec1, const float16_t *&vec2, float16x8_t &acc) {
+    // Load half-precision vectors
+    float16x8_t v1 = vld1q_f16(vec1);
+    float16x8_t v2 = vld1q_f16(vec2);
+    vec1 += 8;
+    vec2 += 8;
+
+    // Calculate differences
+    float16x8_t diff = vsubq_f16(v1, v2);
+    // Square and accumulate
+    acc = vfmaq_f16(acc, diff, diff);
+}
+
+template <unsigned char residual> // 0..31
+float FP16_L2Sqr_NEON_HP(const void *pVect1v, const void *pVect2v, size_t dimension) {
+    const auto *vec1 = static_cast<const float16_t *>(pVect1v);
+    const auto *vec2 = static_cast<const float16_t *>(pVect2v);
+    const auto *const v1End = vec1 + dimension;
+    float16x8_t acc1 = vdupq_n_f16(0.0f);
+    float16x8_t acc2 = vdupq_n_f16(0.0f);
+    float16x8_t acc3 = vdupq_n_f16(0.0f);
+    float16x8_t acc4 = vdupq_n_f16(0.0f);
+
+    // First, handle the partial chunk residual
+    if constexpr (residual % 8) {
+        auto constexpr chunk_residual = residual % 8;
+        // TODO: spacial cases for some residuals and benchmark if its better
+        constexpr uint16x8_t mask = {
+            0xFFFF,
+            (chunk_residual >= 2) ? 0xFFFF : 0,
+            (chunk_residual >= 3) ? 0xFFFF : 0,
+            (chunk_residual >= 4) ? 0xFFFF : 0,
+            (chunk_residual >= 5) ? 0xFFFF : 0,
+            (chunk_residual >= 6) ? 0xFFFF : 0,
+            (chunk_residual >= 7) ? 0xFFFF : 0,
+            0,
+        };
+
+        // Load partial vectors
+        float16x8_t v1 = vld1q_f16(vec1);
+        float16x8_t v2 = vld1q_f16(vec2);
+
+        // Apply mask to both vectors
+        float16x8_t masked_v1 = vreinterpretq_f16_u16(vandq_u16(vreinterpretq_u16_f16(v1), mask));
+        float16x8_t masked_v2 = vreinterpretq_f16_u16(vandq_u16(vreinterpretq_u16_f16(v2), mask));
```

Comment:
Might be worth to benchmark using vbslq instead.
Something like:

float16x8_t zero = vdupq_n_f16(0.0f);
float16x8_t masked_v1 = vbslq_f16(mask, v1, zero);
float16x8_t masked_v2 = vbslq_f16(mask, v2, zero);
acc1 = vfmaq_f16(acc1, masked_v1, masked_v2);

Seems more idiomatic.

Reply:
ran benchmarks [here](https://github.com/RedisAI/VectorSimilarity/actions/runs/14292209421/job/40054886678). It is hard to say if it affects performance, some tests show improvement while others show regression.
I will convert the implantations to the `vbslq_f16` option as it is more readable. Note that for [BF16](https://github.com/RedisAI/VectorSimilarity/pull/623) we don't have this option (unless we reinterpret to another 16-bit type)